login/login.go

```diff
@@ -0,0 +1,384 @@
+	Ga4ghPassportV1   []string `json:"ga4gh_passport_v1"`
+}
+
+type DeviceLogin struct {
+	BaseURL         string
+	ClientID        string
+	S3Target        string
+	PollingInterval int
+	LoginResult     *Result
+	UserInfo        *UserInfo
+	wellKnown       *OIDCWellKnown
+	deviceLogin     *DeviceLoginResponse
+}
+
+type AuthInfo struct {
+	ClientID  string `json:"client_id"`
+	OidcURI   string `json:"oidc_uri"`
+	PublicKey string `json:"public_key"`
+	InboxURI  string `json:"inbox_uri"`
+}
+
+// requests the /info endpoint to fetch the parameters needed for login
+func GetAuthInfo(baseURL string) (*AuthInfo, error) {
+	url := baseURL + "/info"
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	var result AuthInfo
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+// creates a .sda-cli-session file and updates its values
+func (login *DeviceLogin) UpdateConfigFile() error {
+
+	_, err := os.Create(".sda-cli-session")
+	if err != nil {
+		return err
+	}
+	cfg, err := ini.Load(".sda-cli-session")
+	if err != nil {
+		return err
+	}
+
+	s3Config, err := login.GetS3Config()
+	if err != nil {
+		return err
+	}
+
+	err = ini.ReflectFrom(cfg, s3Config)
+	if err != nil {
+		return err
+	}
+	err = cfg.SaveTo(".sda-cli-session")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// NewDeviceLogin() returns a new `DeviceLogin` with the given `url` and
+// `clientID` set.
+func NewDeviceLogin(args []string) (DeviceLogin, error) {
+
+	var url string
+	err := Args.Parse(args[1:])
+	if err != nil {
+		return DeviceLogin{}, errors.New("failed parsing arguments")
+	}
+	if len(Args.Args()) == 1 {
+		url = Args.Args()[0]
+	}
+	log.Println("url: ", url)
+	info, err := GetAuthInfo(url)
+	info.ClientID = "8b7b0168-6b16-4fd2-baec-b0a28b0d5cb0"
+	info.InboxURI = "s3.bp.nbis.se"
+	info.OidcURI = "https://login.elixir-czech.org/oidc"
+	if err != nil {
+		return DeviceLogin{}, errors.New("failed to get auth Info")
+	}
+	log.Println("info: ", info)
+
+	return DeviceLogin{BaseURL: info.OidcURI, ClientID: info.ClientID, PollingInterval: 2, S3Target: info.InboxURI}, nil
+}
+
+// Login() does a full login by fetching the remote configuration, starting the
+// login procedure, and then waiting for the user to complete login.
+func (login *DeviceLogin) Login() error {
+
+	var err error
+	login.wellKnown, err = login.getWellKnown()
+	if err != nil {
+		return fmt.Errorf("failed to fetch .well-known configuration: %v", err)
+	}
+
+	login.deviceLogin, err = login.startDeviceLogin()
+	if err != nil {
+		return fmt.Errorf("failed to start device login: %v", err)
+	}
+	expires := time.Duration(login.deviceLogin.ExpiresIn * int(time.Second))
+	log.Infof("Login started (expires in %v minutes)", expires.Minutes())
+	log.Infof("Go to %v to finish logging in.", login.deviceLogin.VerificationURL)
```

Comment:
That's just a recommendation, in case you think it's better, we could open the default browser directly from the command line with something similar to [this](https://gist.github.com/sevkin/9798d67b2cb9d07cb05f89f14ba682f8)

Reply:
I added it!